FINERACT-38

maektwain opened the pull request:
## Description
I was able to merge the source code which was done @Ippezrobert  on FINERACT-38 , I have done some minimal testing as of now. 

Functional testing needs to be done by @MayaDuruva

Have not written test as of now.

## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [ Y] Commit message starts with the issue number from https://issues.apache.org/jira/projects/FINERACT/. Ex: FINERACT-646 Pockets API.

- [ Y ] Coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions have been followed.

- [ N] API documentation at https://github.com/apache/fineract/blob/develop/api-docs/apiLive.htm has been updated with details of any API changes.

- [ N] Integration tests have been created/updated for verifying the changes made.

- [ Y] All Integrations tests are passing with the new commits.

- [ N] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the list.)

Our guidelines for code reviews is at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide

Commit:
Adding PreviewIncomeAndExpense API Fineract-38

Commit:
Adding PreviewIncomeAndExpense API Fineract-38

Commit:
Adding PreviewIncomeAndExpense API Fineract-38

Commit:
Adding PreviewIncomeAndExpense API Fineract-38

Commit:
Fix in FindByParentID

Commit:
Fix in FindByParentID

Commit:
Fix in FindByParentID

Commit:
Fix in FindByParentID

xurror commented:
This build breaks some checkstyle rules.
Can you check that out?

Commit:
CheckStyle Fixed

Commit:
CheckStyle Fixed

Commit:
CheckStyle Fixed

Commit:
CheckStyle Fixed

maektwain commented:
@xurror Just finished the checkStyleMain task , it seems like the IDE by its own does not picks up the file defined in config , plus it did not fix the files , had to do manually, should be a task where we can fix the files like fixCheckStyleMain 

though my test are still failing locally, but as I see they are passing on travis.

vorburger commented:
> Functional testing needs to be done by @MayaDuruva

@MayaDuruva can you confirm that you will be performing functional testing on this PR?

> Have not written test as of now.

@maektwain Including an integration test for this functionality (in this PR) would be great.

maektwain commented:
This loos like pushed away due to COVID-19, I would like to finish the functional testing and one member is helping me finishing the integration test.

Commit:
Merge pull request #2 from apache/develop

Merge the Apache Fineract - Upstream

singkara committed:
FINERACT-38: Adding 2 test cases to check RunningBalanceNotCalculated… (#1)

* FINERACT-38: Adding 2 test cases to check RunningBalanceNotCalculatedException.

* FINERACT-38 : Renaming and Updating test:

* Fineract-38: Adding 4 test cases signifying the calculation of debits and credits of Income and Expanse accounts.

* FINERACT-38: Making xmlOutput & htmlOutput true for RAT to determine failing libraries

* FINERACT-38: Optimizing imports

* FINERACT-38: Adding Lynx in travis yaml to debug.

* File License Header Added and Some CheckStyleAdded

* Revert "FINERACT-38: Adding Lynx in travis yaml to debug."

This reverts commit f85b05c59a8e7aa847076ee0639e32b8796b7dfa.

* Revert "FINERACT-38: Making xmlOutput & htmlOutput true for RAT to determine failing libraries"

This reverts commit f76f1df1888dbe637a977c32a637013332fdbfc0.

* FINERACT-38: Adding Mockito imports.

* FINERACT-38: Fix in Checkstyle

* FINERACT-38: Correcting checkstyle.

Co-authored-by: Saransh Sharma <marktwain@gmx.us>

singkara committed:
FINERACT-38: Adding 2 test cases to check RunningBalanceNotCalculated… (#1)

* FINERACT-38: Adding 2 test cases to check RunningBalanceNotCalculatedException.

* FINERACT-38 : Renaming and Updating test:

* Fineract-38: Adding 4 test cases signifying the calculation of debits and credits of Income and Expanse accounts.

* FINERACT-38: Making xmlOutput & htmlOutput true for RAT to determine failing libraries

* FINERACT-38: Optimizing imports

* FINERACT-38: Adding Lynx in travis yaml to debug.

* File License Header Added and Some CheckStyleAdded

* Revert "FINERACT-38: Adding Lynx in travis yaml to debug."

This reverts commit f85b05c59a8e7aa847076ee0639e32b8796b7dfa.

* Revert "FINERACT-38: Making xmlOutput & htmlOutput true for RAT to determine failing libraries"

This reverts commit f76f1df1888dbe637a977c32a637013332fdbfc0.

* FINERACT-38: Adding Mockito imports.

* FINERACT-38: Fix in Checkstyle

* FINERACT-38: Correcting checkstyle.

Co-authored-by: Saransh Sharma <marktwain@gmx.us>

singkara committed:
FINERACT-38: Adding 2 test cases to check RunningBalanceNotCalculated… (#1)

* FINERACT-38: Adding 2 test cases to check RunningBalanceNotCalculatedException.

* FINERACT-38 : Renaming and Updating test:

* Fineract-38: Adding 4 test cases signifying the calculation of debits and credits of Income and Expanse accounts.

* FINERACT-38: Making xmlOutput & htmlOutput true for RAT to determine failing libraries

* FINERACT-38: Optimizing imports

* FINERACT-38: Adding Lynx in travis yaml to debug.

* File License Header Added and Some CheckStyleAdded

* Revert "FINERACT-38: Adding Lynx in travis yaml to debug."

This reverts commit f85b05c59a8e7aa847076ee0639e32b8796b7dfa.

* Revert "FINERACT-38: Making xmlOutput & htmlOutput true for RAT to determine failing libraries"

This reverts commit f76f1df1888dbe637a977c32a637013332fdbfc0.

* FINERACT-38: Adding Mockito imports.

* FINERACT-38: Fix in Checkstyle

* FINERACT-38: Correcting checkstyle.

Co-authored-by: Saransh Sharma <marktwain@gmx.us>

singkara committed:
FINERACT-38: Adding 2 test cases to check RunningBalanceNotCalculated… (#1)

* FINERACT-38: Adding 2 test cases to check RunningBalanceNotCalculatedException.

* FINERACT-38 : Renaming and Updating test:

* Fineract-38: Adding 4 test cases signifying the calculation of debits and credits of Income and Expanse accounts.

* FINERACT-38: Making xmlOutput & htmlOutput true for RAT to determine failing libraries

* FINERACT-38: Optimizing imports

* FINERACT-38: Adding Lynx in travis yaml to debug.

* File License Header Added and Some CheckStyleAdded

* Revert "FINERACT-38: Adding Lynx in travis yaml to debug."

This reverts commit f85b05c59a8e7aa847076ee0639e32b8796b7dfa.

* Revert "FINERACT-38: Making xmlOutput & htmlOutput true for RAT to determine failing libraries"

This reverts commit f76f1df1888dbe637a977c32a637013332fdbfc0.

* FINERACT-38: Adding Mockito imports.

* FINERACT-38: Fix in Checkstyle

* FINERACT-38: Correcting checkstyle.

Co-authored-by: Saransh Sharma <marktwain@gmx.us>

maektwain commented:
@vorburger hey , I was able to get some of the tests contributed by one of the GSOC aspirant, maybe down the day we would test the functionality. Could you help speed this up so that we can merge it quickly.

vorburger commented:
@maektwain this PR now has merge conflicts which need to be manually resolve with a rebase...

vorburger committed:
clean-up to remove apparently unused PORT ENV (related to FINERACT-762)

This environment variable is actually never used.  The port of a Tenant
DB instead appears to be read from schemaServerPort of the
tenant_server_connections table.

vorburger committed:
set fineract_tenants_* more nicely in docker-compose.yml

We can use environment variables instead of system properties.

We originally had to use system properties and could not use environment
variables, because these used to read by Tomcat, but since the switch
to Hikari in FINERACT-796 these are read by Spring Boot instead of
Tomcat, so we can now use system properties.

This both "looks nicer" and is useful e.g. for FINERACT-881.

vorburger committed:
add DRIVERCLASS_NAME, PROTOCOL, SUB_PROTOCOL to docker-compose.yml

These are the default values, from jdbc.properties, only added for
clarity and to make it obvious that they need to be changed e.g. for
FINERACT-762.

The different formats and naming conventions are a little bit of a mess,
but changing them now leads to backwards incompatibility, so for now
let's leave as is; possible clean up later in FINERACT-889.

vorburger committed:
add MySQL JDBC Driver (LGPL) to Dockerfile + doc (fixes FINERACT-762)

Note that it is NOT enabled by default, which should be OK for the ASF.

The main point here is just to clearly document it for end-users.

I've obviously tried out what's documented, and can confirm it works.

vorburger committed:
comment out broken MySQL JDBC driver logger (awaiting FINERACT-890)

vorburger committed:
comment out flaky testSavingsAccount_DormancyTracking (FINERACT-852)

thesmallstar committed:
Added One Statement per line and Enforced checkstyle

Commit:
Merge branch 'FINERACT-38' into develop

maektwain commented:
Hello @vorburger I did a quick rebase from develop branch to FINERACT-38 , let me know if this solves the conflict.

vorburger commented:
> Hello @vorburger I did a quick rebase from develop branch to FINERACT-38 , let me know if this solves the conflict.

@maektwain nope it does not... note the "This branch cannot be rebased due to conflicts" message, shown by GitHub just above? Also notice the build failure.. following #750 (based on #740) you have to remove that generics in your new class.

You'll probably want the build to fully to pass locally first after a rebase before you push, and double check yourself that GitHub does not show you "This branch cannot be rebased due to conflicts" anymore.

Commit:
Fixed the AbstractClassLongIssue

Commit:
Fixed the AbstractClassLongIssue

Commit:
Fixed the AbstractClassLongIssue

Commit:
Fixed the AbstractClassLongIssue

maektwain commented:
@vorburger Nopes , it does not show me any conflict issues , quite strange. About the generics I have removed them.
<img width="735" alt="Screenshot 2020-04-15 at 4 11 28 PM" src="https://user-images.githubusercontent.com/8274610/79328545-d7df0b80-7f33-11ea-8df4-c833977e7138.png">

Commit:
Merge branch 'develop' of https://github.com/apache/fineract into develop

Commit:
Merge branch 'FINERACT-38' of https://github.com/muellners/fineract into FINERACT-38

maektwain commented:
Oops, I know what is the issue, I took me while to realise the I was rebasing my own repo so actually this was happening 
Apache Fineract -- My Repo -- all the while I was pulling only My Repo.

Commit:
Merge remote-tracking branch 'origin/FINERACT-38' into FINERACT-38

maektwain commented:
I think, there are many commits now. This will solve all of the issue now.

singkara committed:
FINERACT-38: Removing unused variable.

singkara committed:
FINERACT-38: Removing unused variable.

singkara committed:
FINERACT-38: Removing unused imports

singkara committed:
FINERACT-38: Removing unused imports

singkara committed:
FINERACT-38: Adding ignore statement before failing test case (Not related to this functionality)

singkara committed:
FINERACT-38: Adding ignore statement before failing test case (Not related to this functionality)

maektwain commented:
Have done functional testing also I was able to move and close the accounting closure. Seems to be working fine. 

There is a problem though with SQL Group by query which failed in 5.7 , I fixed it by setting it off , though this is not the bets route to do it.

```
Expression #1 of SELECT list is not in GROUP BY clause and contains nonaggregated column
```

maektwain commented:
ok, I have fixed the above issue.

Commit:
Fixed SQL Query

Commit:
Fixed SQL Query

vorburger commented on fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/ClientSavingsIntegrationTest.java:
```diff
@@ -1907,7 +1908,7 @@ public void testSavingsAccount_WITH_WITHHOLD_TAX_DISABLE_AT_ACCOUNT_LEVEL() {
 
     }
-
+    @Ignore
```
remove this from this PR, and rebase from master, you'll get it from #761 ...

maektwain replied:
@vorburger ack done.

vorburger commented on fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/CalculateIncomeAndExpanseBookingIntegrationTest.java:
```diff
@@ -0,0 +1,167 @@
+        calculateIncomeAndExpenseBooking =  new CalculateIncomeAndExpenseBookingImpl(null, null, null, null, incomeAndExpenseReadPlatformService,officeReadPlatformService);
+     }
+
+    @After
+    public void tearDown() {
+
+    }
```
remove this empty method:

```suggestion
```

maektwain replied:
@singkara could you fix this issue.

vorburger commented on fineract-provider/src/main/java/org/apache/fineract/organisation/office/service/OfficeReadPlatformServiceImpl.java:
```diff
@@ -266,4 +266,20 @@ public OfficeTransactionData retrieveNewOfficeTransactionDetails() {
+        try {
+            return this.jdbcTemplate.queryForList(sqlBuilder.toString(), Long.class,
+                    new Object[] { officeId});
+        } catch (final EmptyResultDataAccessException e) {
+           return null;
+        }
```
This strikes me as curious - is this done like this elsewhere in existing code?

maektwain replied:
@singkara

vorburger reviewed (changes requested):
Had a quick first review, and have 3 minor feedback. Also there was a test failure in `StorageTest` which is most unlikely unrelated to this PR, but a known flaky test, see https://issues.apache.org/jira/browse/FINERACT-828 ... could I suggest that you rebase and we see if it is green?

vorburger committed:
escape all `grouping` in SQL files for MySQL 8 compatibility (#757)

Indirectly related to FINERACT-762 - not sure if strictly required,
but seems like a good idea anyway at this opportunity, so why not.

As originally on https://github.com/apache/fineract/pull/638/files

vorburger committed:
remove JAVA_OPTS from docker-compose.yml (FINERACT-900) (#773)

Commit:
Fixed Merge

maektwain commented:
```
org.apache.fineract.notification.StorageTest > testNotificationStorage FAILED
    org.mockito.exceptions.verification.opentest4j.ArgumentsAreDifferent at StorageTest.java:125
```
Is this supposed to fail

Commit:
Test ASF

vorburger commented:
> StorageTest

@maektwain see https://issues.apache.org/jira/browse/FINERACT-828 about that one (search JIRA!), do you want to see if you can help to fix that one, in a separate PR? The (latest build of this) PR doesn't fail because of that test though, but due to https://issues.apache.org/jira/browse/FINERACT-899... do you know how we could fix that?

Perhaps more importantly than those test failure, this PR has merge conflicts now, and a weird Commit history... 44 commits, of which many which are already on `develop` now - I think something went wrong during your `git rebase`.. do you want to update this to have 1 clean commit only?

maektwain commented:
Hi @singkara , So this has become really big, if you have time maybe you can do this, 

rebase all the things on the latest develop branch , in just a single commit. Could you do this?

maektwain commented:
@vorburger Flaky tests could mean some depending values and arguments are changing, perhaps might need to investigate.

Do we have more information about it?

vorburger commented:
@maektwain all known information is in JIRA, in the tickets linked to https://issues.apache.org/jira/browse/FINERACT-850 - there is one for each known flaky test. Exactly WHY each of those are flaky is what working on those JIRAs requires to find out... :smile_cat:

singkara committed:
FINERACT-38: Should not be return null, but shall be OfficeNotFoundDexeption.

singkara committed:
FINERACT-38: Rebase

singkara committed:
FINERACT: Removing unused import

maektwain commented:
@singkara could you do this , we close this PR and we might need to just add this commit.

7f9967a
142da4f
2fa66ce
19783b1
091f186

That is all, and add your own tests also and replay it on the our repo before we do open a PR here, and we try to do a squash in single commit.

Let me know if you have any question.

singkara commented:
Hi, @maektwain I have created the new branch FINERACT-38_rebase for the new changes. However, while cherrypicking, I had merge conflicts in multiple files and since I am not the writer, I am not sure which ones to keep. Perhaps we could get on a call and finish this.
The attached screenshot shows the conflicts just because of cherrypicking 1st commit.
<img width="1374" alt="Screen Shot 2020-04-19 at 19 37 33" src="https://user-images.githubusercontent.com/44347219/79703108-a5a51500-8277-11ea-927f-f968c8fc9df3.png">
Although I have an intuition of which changes shall go in, just to be cautious I would ask you to review the merge conflicts once.

maektwain commented:
Take the head one, and try to replay and lets close this off.

vorburger commented:
@maektwain @singkara see (new) https://github.com/apache/fineract#pull-requests and do not wait for an active maintainer to come and help you re. build failures - it's your responsibility to get a green build. Have you checked the log, and done what the README suggests to do in such cases? :smiley_cat:
